src/Frontend/Core/Engine/Url.php

```diff
@@ -232,27 +232,27 @@ private function determineLanguage(string $queryString): string
         $chunks = (array) explode('/', $queryString);
 
         // the language is present in the URL
-        if (isset($chunks[0]) && in_array($chunks[0], $possibleLanguages)) {
-            // define language
-            $language = (string) $chunks[0];
-            $this->setLanguageCookie($language);
-
-            Model::getSession()->set('frontend_language', $language);
-
-            return $language;
-        }
+        if (!isset($chunks[0]) || !in_array($chunks[0], $possibleLanguages, true)) {
+            $cookie = $this->getContainer()->get('fork.cookie');
+            if ($cookie->has('frontend_language')
+                && in_array($cookie->get('frontend_language'), $redirectLanguages, true)
+            ) {
+                $this->redirectToLanguage($cookie->get('frontend_language'));
+            }
 
-        $cookie = $this->getContainer()->get('fork.cookie');
-        if ($cookie->has('frontend_language')
-            && in_array($cookie->get('frontend_language'), $redirectLanguages, true)
-        ) {
-            $this->redirectToLanguage($cookie->get('frontend_language'));
+            // default browser language
+            $language = Language::getBrowserLanguage();
+            $this->setLanguageCookie($language);
+            $this->redirectToLanguage($language);
         }
 
-        // default browser language
-        $language = Language::getBrowserLanguage();
+        // define language
+        $language = (string) $chunks[0];
```

Comment:
Maybe you're forgetting a return statement in the previous if statement? Otherwise `$chunks[0]` can still be empty here.

Reply:
Forgot to revert this. The change was part of making phpstan understand our redirects but instead I confused you :p